File: php/logbrew-php/src/LogBrewHttpRequestTelemetry.php
<?php

declare(strict_types=1);

namespace LogBrew;

/**
 * Explicit app-owned HTTP request telemetry with trace/log/error correlation.
 *
 * @phpstan-import-type Metadata from LogBrewClient
 */
final class LogBrewHttpRequestTelemetry
{
    private readonly int $startedAtNs;
    private bool $finished = false;

    private function __construct(
        private readonly LogBrewClient $client,
        public readonly string $method,
        public readonly string $routeTemplate,
        public readonly LogBrewTraceContext $trace
    ) {
        $this->startedAtNs = hrtime(true);
    }

    /**
     * Start request telemetry, continuing a valid incoming traceparent or falling back to a local root trace.
     */
    public static function start(
        LogBrewClient $client,
        string $method,
        string $routeTemplate,
        ?string $incomingTraceparent = null
    ): self {
        return new self(
            $client,
            self::normalizeMethod($method),
            self::normalizeRouteTemplate($routeTemplate),
            LogBrewTraceContext::fromIncomingTraceparentOrCreateRoot($incomingTraceparent)
        );
    }

    /**
     * Start request telemetry from an app-created trace context.
     */
    public static function startWithTraceContext(
        LogBrewClient $client,
        string $method,
        string $routeTemplate,
        LogBrewTraceContext $trace
    ): self {
        return new self($client, self::normalizeMethod($method), self::normalizeRouteTemplate($routeTemplate), $trace);
    }

    /**
     * Activate this request trace while handler code runs.
     */
    public function activate(): LogBrewTraceScope
    {
        return LogBrewTrace::activate($this->trace);
    }

    /**
     * Return explicit outgoing propagation headers for app-owned HTTP clients.
     *
     * @return array{traceparent:string}
     */
    public function outgoingHeaders(): array
    {
        return $this->trace->headers();
    }

    /**
     * Finish the request span with route/status metadata.
     *
     * @param array<string, mixed> $metadata
     */
    public function finishSpan(string $eventId, string $timestamp, int $statusCode, array $metadata = []): void
    {
        $this->finish($eventId, null, $timestamp, $statusCode, $metadata);
    }

    /**
     * Finish the request span and matching http.server.duration metric.
     *
     * @param array<string, mixed> $metadata
     */
    public function finishSpanAndMetric(
        string $spanEventId,
        string $metricEventId,
        string $timestamp,
        int $statusCode,
        array $metadata = []
    ): void {
        $this->finish($spanEventId, $metricEventId, $timestamp, $statusCode, $metadata);
    }

    /**
     * @param array<string, mixed> $metadata
     */
    private function finish(
        string $spanEventId,
        ?string $metricEventId,
        string $timestamp,
        int $statusCode,
        array $metadata
    ): void {
        if ($this->finished) {
            throw new SdkError('validation_error', 'HTTP request telemetry is already finished');
        }
        self::requireStatusCode($statusCode);

        $durationMs = $this->elapsedMilliseconds();
        $requestMetadata = LogBrewTrace::metadataWithTrace($this->trace, $this->requestMetadata($statusCode, $metadata));
        $spanAttributes = [
            'name' => sprintf('%s %s', $this->method, $this->routeTemplate),
            'traceId' => $this->trace->traceId,
            'spanId' => $this->trace->spanId,
            'status' => self::statusFromHttpStatus($statusCode),
            'durationMs' => $durationMs,
            'metadata' => $requestMetadata,
        ];
        if ($this->trace->parentSpanId !== null) {
            $spanAttributes['parentSpanId'] = $this->trace->parentSpanId;
        }

        $this->client->span($spanEventId, $timestamp, $spanAttributes);
        if ($metricEventId !== null) {
            $this->client->metric($metricEventId, $timestamp, [
                'name' => 'http.server.duration',
                'kind' => 'histogram',
                'value' => $durationMs,
                'unit' => 'ms',
                'temporality' => 'delta',
                'metadata' => $requestMetadata,
            ]);
        }

        $this->finished = true;
    }

    private function elapsedMilliseconds(): float
    {
        return max(0.0, (hrtime(true) - $this->startedAtNs) / 1_000_000);
    }

    /**
     * @param array<string, mixed> $metadata
     * @return Metadata
     */
    private function requestMetadata(int $statusCode, array $metadata): array
    {
        $copied = LogBrewClient::copyPrimitiveMetadata($metadata);
        $copied['method'] = $this->method;
        $copied['routeTemplate'] = $this->routeTemplate;
        $copied['statusCode'] = $statusCode;
        return $copied;
    }

    private static function normalizeMethod(string $method): string
    {
        LogBrewClient::requireNonEmpty('HTTP request method', $method);
        $normalized = strtoupper(trim($method));
        if (preg_match('/^[A-Z]+$/', $normalized) !== 1) {
            throw new SdkError('validation_error', 'HTTP request method must be a valid HTTP method');
        }

        return $normalized;
    }

    private static function normalizeRouteTemplate(string $routeTemplate): string
    {
        LogBrewClient::requireNonEmpty('HTTP request routeTemplate', $routeTemplate);
        $trimmed = trim($routeTemplate);
        $parts = parse_url($trimmed);
        if (is_array($parts) && isset($parts['scheme'], $parts['host'])) {
            $path = (string) ($parts['path'] ?? '/');
            return $path === '' ? '/' : $path;
        }

        $cutoff = self::firstPresentIndex(strpos($trimmed, '?'), strpos($trimmed, '#'));
        if ($cutoff !== null) {
            $trimmed = rtrim(substr($trimmed, 0, $cutoff));
        }
        if ($trimmed === '') {
            throw new SdkError('validation_error', 'HTTP request routeTemplate must be non-empty');
        }

        return $trimmed;
    }

    private static function requireStatusCode(int $statusCode): void
    {
        if ($statusCode < 100 || $statusCode > 599) {
            throw new SdkError('validation_error', 'HTTP request statusCode must be between 100 and 599');
        }
    }

    /** @return 'ok'|'error' */
    private static function statusFromHttpStatus(int $statusCode): string
    {
        return $statusCode >= 500 ? 'error' : 'ok';
    }

    private static function firstPresentIndex(int|false $first, int|false $second): ?int
    {
        if ($first === false) {
            return $second === false ? null : $second;
        }
        if ($second === false) {
            return $first;
        }

        return min($first, $second);
    }
}

File: java/logbrew-java/src/main/java/co/logbrew/sdk/LogBrewHttpRequestTelemetry.java
package co.logbrew.sdk;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.regex.Pattern;

/**
 * Dependency-free helper for app-owned HTTP server request spans.
 *
 * <p>The helper records a request span and optional {@code http.server.duration}
 * metric with the same trace/span IDs that logger integrations read from the
 * active trace scope.</p>
 */
public final class LogBrewHttpRequestTelemetry {
    private static final Pattern METHOD_PATTERN = Pattern.compile("^[A-Za-z]+$");

    private final LogBrewClient client;
    private final String method;
    private final String routeTemplate;
    private final LogBrewTraceContext traceContext;
    private final Map<String, Object> metadata;
    private final long startedNanos;
    private boolean finished;

    private LogBrewHttpRequestTelemetry(
        LogBrewClient client,
        String method,
        String routeTemplate,
        LogBrewTraceContext traceContext,
        Map<String, ?> metadata
    ) {
        this.client = Objects.requireNonNull(client, "client");
        this.method = normalizeMethod(method);
        this.routeTemplate = sanitizeRouteTemplate(routeTemplate);
        this.traceContext = Objects.requireNonNull(traceContext, "traceContext");
        Map<String, Object> copiedMetadata = Validation.copyMetadata(metadata);
        this.metadata = copiedMetadata == null ? new LinkedHashMap<>() : copiedMetadata;
        this.startedNanos = System.nanoTime();
    }

    /**
     * Starts request telemetry with a generated local trace.
     */
    public static LogBrewHttpRequestTelemetry start(LogBrewClient client, String method, String routeTemplate) {
        return start(client, method, routeTemplate, LogBrewTraceContext.generate(), null);
    }

    /**
     * Starts request telemetry by continuing an incoming W3C traceparent.
     */
    public static LogBrewHttpRequestTelemetry start(
        LogBrewClient client,
        String method,
        String routeTemplate,
        String traceparent
    ) {
        return start(client, method, routeTemplate, traceContextFromIncomingHeader(traceparent), null);
    }

    /**
     * Starts request telemetry by continuing an incoming W3C traceparent with base metadata.
     */
    public static LogBrewHttpRequestTelemetry start(
        LogBrewClient client,
        String method,
        String routeTemplate,
        String traceparent,
        Map<String, ?> metadata
    ) {
        return start(client, method, routeTemplate, traceContextFromIncomingHeader(traceparent), metadata);
    }

    /**
     * Starts request telemetry from an explicit LogBrew trace context.
     */
    public static LogBrewHttpRequestTelemetry start(
        LogBrewClient client,
        String method,
        String routeTemplate,
        LogBrewTraceContext traceContext
    ) {
        return start(client, method, routeTemplate, traceContext, null);
    }

    /**
     * Starts request telemetry from an explicit LogBrew trace context and base metadata.
     */
    public static LogBrewHttpRequestTelemetry start(
        LogBrewClient client,
        String method,
        String routeTemplate,
        LogBrewTraceContext traceContext,
        Map<String, ?> metadata
    ) {
        return new LogBrewHttpRequestTelemetry(client, method, routeTemplate, traceContext, metadata);
    }

    /**
     * Returns the request trace context.
     */
    public LogBrewTraceContext traceContext() {
        return traceContext;
    }

    /**
     * Makes the request trace active for logs, issues, and explicit metrics.
     */
    public LogBrewTrace.Scope activate() {
        return LogBrewTrace.activate(traceContext);
    }

    /**
     * Returns an outbound carrier containing only W3C {@code traceparent}.
     */
    public Map<String, String> outgoingHeaders() {
        return traceContext.headers();
    }

    /**
     * Records a request span with the elapsed duration.
     */
    public void finishSpan(String eventId, String timestamp, int statusCode) {
        finishSpan(eventId, timestamp, statusCode, elapsedMs());
    }

    /**
     * Records a request span with an explicit duration.
     */
    public void finishSpan(String eventId, String timestamp, int statusCode, double durationMs) {
        ensureNotFinished();
        client.span(eventId, timestamp, spanAttributes(statusCode, durationMs));
        finished = true;
    }

    /**
     * Records a request span and {@code http.server.duration} metric with the same elapsed duration.
     */
    public void finishSpanAndMetric(String spanEventId, String metricEventId, String timestamp, int statusCode) {
        finishSpanAndMetric(spanEventId, metricEventId, timestamp, statusCode, elapsedMs());
    }

    /**
     * Records a request span and {@code http.server.duration} metric with an explicit duration.
     */
    public void finishSpanAndMetric(
        String spanEventId,
        String metricEventId,
        String timestamp,
        int statusCode,
        double durationMs
    ) {
        ensureNotFinished();
        client.span(spanEventId, timestamp, spanAttributes(statusCode, durationMs));
        client.metric(metricEventId, timestamp, metricAttributes(statusCode, durationMs));
        finished = true;
    }

    private SpanAttributes spanAttributes(int statusCode, double durationMs) {
        validateStatusCode(statusCode);
        SpanAttributes attributes = SpanAttributes
            .create(method + " " + routeTemplate, traceContext.traceId(), traceContext.spanId(), spanStatus(statusCode))
            .durationMs(durationMs)
            .metadata(requestMetadata(statusCode));
        if (traceContext.parentSpanId() != null) {
            attributes.parentSpanId(traceContext.parentSpanId());
        }
        return attributes;
    }

    private MetricAttributes metricAttributes(int statusCode, double durationMs) {
        validateStatusCode(statusCode);
        return MetricAttributes
            .create("http.server.duration", "histogram", durationMs, "ms", "delta")
            .metadata(requestMetadata(statusCode));
    }

    private Map<String, Object> requestMetadata(int statusCode) {
        Map<String, Object> values = new LinkedHashMap<>(metadata);
        values.put("method", method);
        values.put("routeTemplate", routeTemplate);
        values.put("statusCode", Integer.valueOf(statusCode));
        return LogBrewTrace.metadataWithTrace(traceContext, values);
    }

    private double elapsedMs() {
        return (System.nanoTime() - startedNanos) / 1_000_000.0;
    }

    private void ensureNotFinished() {
        if (finished) {
            throw new SdkException("validation_error", "request telemetry has already been finished");
        }
    }

    private static String normalizeMethod(String method) {
        Validation.requireNonEmpty("HTTP method", method);
        String normalized = method.trim().toUpperCase(java.util.Locale.ROOT);
        if (!METHOD_PATTERN.matcher(normalized).matches()) {
            throw new SdkException("validation_error", "HTTP method must contain letters only");
        }
        return normalized;
    }

    private static String sanitizeRouteTemplate(String routeTemplate) {
        Validation.requireNonEmpty("routeTemplate", routeTemplate);
        String trimmed = routeTemplate.trim();
        int schemeIndex = trimmed.indexOf("://");
        if (schemeIndex >= 0) {
            int pathStart = trimmed.indexOf('/', schemeIndex + 3);
            trimmed = pathStart >= 0 ? trimmed.substring(pathStart) : "/";
        }
        int queryIndex = trimmed.indexOf('?');
        int hashIndex = trimmed.indexOf('#');
        int end = trimmed.length();
        if (queryIndex >= 0) {
            end = Math.min(end, queryIndex);
        }
        if (hashIndex >= 0) {
            end = Math.min(end, hashIndex);
        }
        String sanitized = trimmed.substring(0, end).trim();
        if (sanitized.isEmpty()) {
            throw new SdkException("validation_error", "routeTemplate must be non-empty after sanitization");
        }
        return sanitized;
    }

    private static void validateStatusCode(int statusCode) {
        if (statusCode < 100 || statusCode > 599) {
            throw new SdkException("validation_error", "HTTP statusCode must be an integer from 100 to 599");
        }
    }

    private static String spanStatus(int statusCode) {
        return statusCode >= 500 ? "error" : "ok";
    }

    private static LogBrewTraceContext traceContextFromIncomingHeader(String traceparent) {
        if (traceparent == null || traceparent.trim().isEmpty()) {
            return LogBrewTraceContext.generate();
        }
        try {
            return LogBrewTraceContext.fromTraceparent(traceparent);
        } catch (SdkException error) {
            return LogBrewTraceContext.generate();
        }
    }
}
